File: src/lib/code/mcs-216/section-1/session-2/knapsack.py
# Fractional Knapsack in Python
def get_max_value(profits, weights, capacity):
    # Calculate profit/weight ratio and keep track of original index
    items = []
    for i in range(len(profits)):
        items.append(
            {
                "id": i + 1,
                "profit": profits[i],
                "weight": weights[i],
                "ratio": profits[i] / weights[i],
            }
        )

    # Sort items by ratio in descending order (Greedy Strategy)
    items.sort(key=lambda x: x["ratio"], reverse=True)

    total_value = 0.0
    for item in items:
        if capacity > 0 and capacity >= item["weight"]:
            # Take the full item
            capacity -= item["weight"]
            total_value += item["profit"]
        elif capacity > 0:
            # Take a fraction of the item to fill remaining capacity
            total_value += item["profit"] * (capacity / item["weight"])
            capacity = 0
            break

    return total_value


# Solving Question 1
p1, w1, c1 = [15, 5, 20, 8, 7, 20, 6], [3, 4, 6, 8, 2, 2, 3], 18
print(f"Result Q1: {get_max_value(p1, w1, c1):.2f}")

File: src/lib/code/mcs-216/section-1/session-2/knapsack.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int id;
    double p, w, ratio;
} Item;

// Comparator for descending order based on ratio
int compareItems(const void* a, const void* b) {
    Item* itemA = (Item*)a;
    Item* itemB = (Item*)b;
    if (itemB->ratio > itemA->ratio) return 1;
    if (itemB->ratio < itemA->ratio) return -1;
    return 0;
}

double fractionalKnapsack(double p[], double w[], int n, double capacity) {
    Item items[n];
    for (int i = 0; i < n; i++) {
        items[i].id = i + 1;
        items[i].p = p[i];
        items[i].w = w[i];
        items[i].ratio = p[i] / w[i];
    }

    qsort(items, n, sizeof(Item), compareItems);

    double totalProfit = 0.0;
    for (int i = 0; i < n; i++) {
        if (capacity >= items[i].w) {
            capacity -= items[i].w;
            totalProfit += items[i].p;
        } else {
            // Add the fractional part
            totalProfit += items[i].p * (capacity / items[i].w);
            break;
        }
    }
    return totalProfit;
}

int main() {
    double p2[] = {20, 30, 40, 32, 55};
    double w2[] = {5, 8, 10, 12, 15};
    printf("Result Q2: %.2f\n", fractionalKnapsack(p2, w2, 5, 50));
    return 0;
}
